### PR TITLE
Clickhouse cluster reconfigurator execution

### PR DESCRIPTION
**Pull request by andrewjstone**
Create a clickhouse-admin-client and use it in reconfigurator execution for clickhouse keepers and servers.

Note that since the policy is disabled, there won't be a `Some(clickhouse_cluster_config)` in the blueprint and the execution will be skipped.

### CODE REVIEW

File: nexus/reconfigurator/execution/src/clickhouse.rs [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+                );
+                return Err(vec![e]);
+            }
+        };
+
+    let mut errors = vec![];
+
+    // Inform each clickhouse-admin server in a keeper zone about its keeper's
+    // configuration
+    for config in keeper_configs {
```
Comment (andrewjstone) [inline nexus/reconfigurator/execution/src/clickhouse.rs:80]: Should almost certainly do theses client ops in parallel.

File: nexus/reconfigurator/execution/src/clickhouse.rs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+                    "to clickhouse-admin; admin_url = {}, error = {}",
+                ),
+                config.settings.id,
+                admin_url,
+                e
+            ));
+        }
+    }
+
+    // Inform each clickhouse server about its configuration
```
Comment (andrewjstone) [inline nexus/reconfigurator/execution/src/clickhouse.rs:108]: Fix this comment

File: nexus/reconfigurator/execution/src/clickhouse.rs [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+        .collect();
+
+    let mut remote_servers =
+        Vec::with_capacity(clickhouse_cluster_config.servers.len());
+
+    for (zone_id, server_id) in &clickhouse_cluster_config.servers {
+        remote_servers.push(ClickhouseHost::Ipv6(
+            *server_ips.get(zone_id).ok_or_else(|| {
+                anyhow!(
+                    "Failed to retrieve zone {} for server id {}",
```
Comment (karencfv) [inline nexus/reconfigurator/execution/src/clickhouse.rs:188]: Optional nit:
```suggestion
                    "Failed to retrieve zone {} for replica server id {}",
```

**Review by karencfv — APPROVED** [review]
Just a bunch of optional nits for error messages, and a small change, but everything else looks good to me.

Almost there!!! 🎉

File: nexus/reconfigurator/execution/src/clickhouse.rs [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+
+    let server_configs =
+        match server_configs(zones, clickhouse_cluster_config, keeper_hosts) {
+            Ok(server_configs) => server_configs,
+            Err(e) => {
+                // We can't proceed if we fail to generate configs.
+                // Let's be noisy about it.
+                error!(
+                    opctx.log,
+                    "Failed to generate clickhouse server configs: {e}"
```
[the remark this answers is not in the document]
Comment (karencfv) [reply nexus/reconfigurator/execution/src/clickhouse.rs:73]: Sometimes keepers are called "keeper servers" in the ClickHouse docs, so perhaps best to be explicit?

File: nexus/reconfigurator/execution/src/clickhouse.rs (hunk 1 above)
Comment (andrewjstone) [reply nexus/reconfigurator/execution/src/clickhouse.rs:85]: done.

File: nexus/reconfigurator/execution/src/clickhouse.rs [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+                return Err(vec![e]);
+            }
+        };
+
+    let mut errors = vec![];
+    let log = opctx.log.clone();
+
+    // Inform each clickhouse-admin server in a keeper zone or server zone about
+    // its node's configuration
+    let mut futs = FuturesUnordered::new();
```
Comment (sunshowers) [inline nexus/reconfigurator/execution/src/clickhouse.rs:84]: Only thing I'd add here is to consider restricting concurrency via `buffer_unordered`.

**Review by sunshowers — APPROVED** [review]
Looks good, thanks!

File: nexus/reconfigurator/execution/src/clickhouse.rs [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+) -> Result<(), Vec<anyhow::Error>> {
+    let keeper_configs = match keeper_configs(zones, clickhouse_cluster_config)
+    {
+        Ok(keeper_configs) => keeper_configs,
+        Err(e) => {
+            // We can't proceed if we fail to generate configs.
+            // Let's be noisy about it.
+            error!(
+                opctx.log,
+                "failed to generate clickhouse keeper configs: {e}"
```
[the remark this answers is not in the document]
Comment (andrewjstone) [reply nexus/reconfigurator/execution/src/clickhouse.rs:54]: I prefer to keep things short and terse for single line log messages.

File: nexus/reconfigurator/execution/src/clickhouse.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply nexus/reconfigurator/execution/src/clickhouse.rs:73]: These are consistent names used throughout the code and I prefer to keep things short in logs. The clickhouse command itself uses `clickhouse server` and `clickhouse keeper`

File: nexus/reconfigurator/execution/src/clickhouse.rs [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+        ));
+        let admin_url = format!("http://{admin_addr}");
+        let log = log.new(slog::o!("admin_url" => admin_url.clone()));
+        futs.push(Either::Left(async move {
+            let client = Client::new(&admin_url, log.clone());
+            client.generate_keeper_config(&config).await.map(|_| ()).map_err(
+                |e| {
+                    anyhow!(
+                        concat!(
+                            "failed to send config for clickhouse keeper ",
```
[the remark this answers is not in the document]
Comment (andrewjstone) [reply nexus/reconfigurator/execution/src/clickhouse.rs:100]: same as above

**Comment by andrewjstone** [comment]
Thanks for the quick reviews @karencfv @sunshowers. Much appreciated!

**Pull request merged**